### ISSUE TITLE
Cache a node: GL.Node saveBuffer+cached props

### ISSUE THREAD

**Issue by gre** [gre/gl-react#27]
We should be able to **cache a node** (which means part of the rendering tree) **using a previously saved buffer**. We should also be able to **pipe the previous framebuffer as a texture** for more advanced use-cases like cellular automaton or simply persistence effects _(and making gl-react more compatible with shaders from shadertoy.com and glslsandbox.com)_. This can also be used in combination of `discard` in GLSL to re-render only part of a buffer (preserve part of the previous buffer) to allow interesting things like drawing a brush (see https://github.com/ProjectSeptemberInc/gl-react-native/issues/19).
## API Draft
- `<GL.Node shouldUpdate={false} saveBuffer="foo">`: a way to cache node's render computation.
  - **`saveBuffer`**: keep the rendering content for the next frame (using a framebuffer). It's a bit like 'key' semantic in React but it's global across the whole tree.
  - **`shouldUpdate`**: if set to `true` (default) will make the `GL.Node` render normally. if set to `false`, will shortcut the rendering to re-use the previous framebuffer instead.
- special symbols that can be passed to `uniforms`:
  - **`GL.PreviousBuffer`**: the previous rendered buffer.
  - **`GL.HasPreviousBuffer`**: boolean to know if `GL.PreviousBuffer` is provided _(if false, PreviousBuffer will be rendered black)_.
## Implementation

`gl-react-native` and `gl-react-dom` implementations will have to use a new framebuffer collection that is preserved after each render. This new collection of framebuffers will be solved by `gl-react`. There is 2 categories: **"shared" fbos** (if supported by the implementation (e.g iOS), fbos are shared between the surface instances) and **"preserved" fbos** that exclusively belong to each surface instance. The preserved fbos are identified by an id (user provided in saveBuffer props). _(so `preservedFbos` will be a Map / Dictionary / JSObject from that id to the fbo object)_.
We need a new internal format to describe a fbo: `{ type: "shared", id: 1 }` & `{ type: "preserved", id: "foo" }`.
Everywhere we used a fboId, it should adopt this new format (e.g in texture types).
### in draw()

in draw() method of implementations we need to do this:
- save glcanvas' preservedFbos into previousPreservedFbos.
- create a newPreservedFbos.
- every time a uniform references a preservedFbos, we pick it from the previousPreservedFbos map.
- ~~every time a node uses saveBuffer, we still use the shared fboId to store the rendering, but at the end of that rendering, we store that shared fboId into newPreservedFbos and create a new fbo in the shared fboId (this is the safest, but we should try to limit this recreation.., i'm already thinking of a smarter way but that might be premature optim).~~ – I need to rethink that, i'm not 100% sure if this works and if it's the simpler way
- set newPreservedFbos into glcanvas' preservedFbos for next draw().
## gl-react

there must be an invariant that check there are no dups in saveBuffer in the tree.

> This feature would have to be carefully used because storing part of the tree cost one more framebuffer to keep in memory (for each view instance) but in some cases we might want to cache a complex rendering that rarely changes. It's a tradeoff between GPU computation and Memory.

---

~~`<GL.Node shader={...} saveBuffer="foo" cached={true || false} />~~
-  ~~`saveBuffer` because I think we need a unique id to locate the node if tree changes a lot. this also makes it consistent with #25~~
-  ~~`cached` is a boolean to indicate the lib should use a previously saved buffer.~~

---

~~also, saved buffer can be used in uniforms:~~

``` html
<!--
<GL.Node shader={...}
  uniforms={{
    backbuffer: GL.fromBuffer("foo")
  }}
/>
-->
```

~~`GL.fromBuffer` references a buffer saved with `saveBuffer` prop in the last rendering frame, it can be the node itself (allows to do persistence effect – in that case, implementation will need to use a trick either to switch between 2 fbos or copy fbos one into another).~~

**Comment by gre** [comment]
more thoughts on this

actually I'm not sure we even need fromBuffer, GL.Node saveBuffer and cached might be enough because of the work done on "factorizing" tree.

fromBuffer is probably an anti-pattern and we would do instead this equivalent feature by passing the GL.Node directly:

```
const cachePartOfTheTree = <GL.Node shader={...}
  saveBuffer="foo"
  cached={true || false}
/>;

<GL.Node shader={...}
  uniforms={{
    cachePartOfTheTree
  }}
/>
```

**Comment by gre** [comment]
i'm unscheduling this for next milestone, because I think it's not mature yet, need to give a bit more thought on maybe simplifying this API

**Comment by gre** [comment]
Latest thoughts:

- `<GL.Node shouldUpdate={false} saveBuffer="foo">`: a way to cache node's render computation.
  - **`saveBuffer`**: keep the rendering content for the next frame (using a framebuffer).
  - **`shouldUpdate`**: if set to `true` (default) will make the `GL.Node` render normally. if set to `false`, will shortcut the rendering to re-use the previous framebuffer instead.
- special symbols that can be passed to `uniforms`:
  - **`GL.PreviousBuffer`**: the previous rendered buffer.
  - **`GL.HasPreviousBuffer`**: boolean to know if `GL.PreviousBuffer` is provided *(if false, PreviousBuffer is black)*.

**Comment by gre** [comment]
`<GL.Node saveBuffer="foo">` will a good use-case to allow the use of `discard` https://github.com/ProjectSeptemberInc/gl-react-native/issues/19

**Comment by gre** [comment]
maybe worth introducing a new decorator `GL.Cache` instead of adding more into GL.Node. The thing is maybe we would want to factorize some of this mecanism for contents too ( #53 & #56 ).

```
<GL.Cache shouldUpdate={false} saveBuffer="foo">
   ...
</GL.Cache>
```

**Comment by gre** [comment]
the GL.Cache idea is not that compatible with GL.PreviousBuffer simplicity, so we might only introduce GL.Cache for content?
this issue is hard to tackle.

**Comment by morukutsu** [comment]
Hi, I was interested by the "previous frame feature" described in the first post (to implement motion blur / accumulation like effects). Using Surface.captureFrame was too slow for real time as expected.

I implemented a quick demo of the feature here: https://github.com/morukutsu/gl-react-dom/commit/4b6705eeedfec67585d4eda35b40231fef7e52b4

There are still some issues:
- implementation is hacky because of the magic number I use for the previous frame fbo id
- previousFrame is a hardcoded name which triggers the feature, using the "GL.PreviousBuffer" semantic would be better
- I didn't check the implementation for the native version

What are your plans regarding this feature? I guess it is not really useful for most use cases but I kinda like it! And it's not too hard to implement properly.

**Comment by SereznoKot** [comment]
@morukutsu I'm really interested in this feature as well, your solution seems to work, do you have any plans to refine it and make a PR?

**Comment by morukutsu** [comment]
@SereznoKot hi, yep I have some plans.
I used my workaround in my app and I had some issues: it's only possible to get the last frame of the whole surface, not per GL.Node. In my project I have a complex stack of effects and I would like to have local feedback effects.

So I'm thinking about something in the lines of what @gre suggested:

`
<GL.Node shouldUpdate={false} saveBuffer="foo">
...
// to retrieve the texture as a uniform
GL.fromBuffer("foo")
`

I went through the source code some weeks ago and it doesn't seem too hard to implement. The current problem is that I'm not too familiar with the native part (gl-react-native, which has both Android and iOS low level implementations). So if I start working on it it would be for the gl-react-dom first.

**Comment by SereznoKot** [comment]
@morukutsu cool! Let me know, if you need any help with that (not sure, if I can help with native, more likely with dom)

**Comment by gre** [comment]
as side notes to this, I'm not satisfied enough with how gl-react is implemented today to easily move forward to this more complex step (the fact gl-react handle the tree of VDOM itself, unfold the tree, etc.. it's quite complex code already). I'm considering reimplementing it inspiring from how `react-art` and `react-music` is implemented.

I can't guarantee when this will come but I want to make a quick PoC to see if it's feasible, and figure out what tradeoffs will be made:

- the idea of sharing VDOM to dedup the shared rendering will probably be removed and instead, user will have to define what is shared, maybe a bit like in react-music *bus* idea. React knows how to handle tree of VDOM, but you can't represent graphs in that paradigm.

```js
// idea
<Surface>
	<HelloTexture>
		<GL.Shared from="foo" />
	</HelloTexture>
	<GL.Shared to="foo">
		<Blur>{...}</Blur>
	</GL.Shared>
</Surface>
```
- basic idea is truely benefiting React diffing, so we have less code in gl-react, probably better perf, and for sure cache will be easier to implement. Imagine if a GLNode is a host component (dom/native) that hold things like the Framebuffer. We could easily just introduce a snapshot method that directly hit that framebuffer and gets its pixels.
Also, if doing this more React-y approach,`shouldComponentUpdate` should just be enough to implement cache :)
Imagine if you can also have state in part of your stack of effects!

lot of possibilities, but I need to see how to make this possible, as a GL.Node is not something on top of DOM or UIView but it's a shadow abstract thing for working with GL, and the GLSurface is the true renderer at the end.

thoughts?

**Comment by morukutsu** [comment]
I had a look at the VDOM dedupe feature. It's very useful but as you said, it is a very complicated piece of code. Also, I'm wondering if most of the use cases does not really takes advantage of this feature. 
If so, it make sense to remove this code in favor of a simpler implementation and let the user manage his own caching mechanism.

> We could easily just introduce a snapshot method that directly hit that framebuffer and gets its pixels.
If there is an active OpenGL context during the GL.Node render (which is not the case at the moment). The GL.Node can be rendered to a FBO. Then, this FBO is passed to the parent (another GL.Node or a Surface for final display). The problem is that we need one surface sized FBO for each GL.Node in the tree if caching is enabled. If not, old FBOs can be reused. 

Basically, the Surface rendering code is moved to the GL.Node class. Surface would just manage the OpenGL context and pass it to the children.

> One subtle but hard problem is how to handle the rasterized content?
I'm not sure I understand everything about this problem. React ref are a nice mechanism, however, assuming rasterized content is rendered to a FBO. Using a helper to retrieve FBOs from a global state is not bad either (the GL.fromBuffer("foo") thing ).

**Comment by gre** [comment]
Sorry to have been a bit too talkative my 5-days-ago comment, but basically most issues I was talking about are fixed in a PoC implementation I've started this weekend ( see also https://github.com/ProjectSeptemberInc/gl-react/issues/70#issuecomment-254172991 ). Still a WIP, I'm exploring different things and still change the API a lot, so I won't publish this right now, but soon I hope ;) I will also have to see how I can port this to native, and how I can reuse the maximum code on JS side (which also might change some API).

Basically I do what you said: less magic in the lib and more control to user, in short: dropping the dedupe logic, embracing React lifecycle.

I'm using the context to connect things together. 

One "hack" is that GL.Node render() method is like: `<span>{all uniform sub nodes}</span>`, so it's rendered normally in React. I also even put that inside the `<canvas>` tag (it can have content, it's just not visible. well, perfect!!)

I attach one framebuffer to each GL.Node instance. (except for the "back buffering", you just set a `backbuffering` prop, and you have 2 fbos that get swaped each draw time)

*I have not yet thought about sharing and reusing fbos, attaching one fbo per GL.Node makes things so much simpler and allow to do things like caching (redraw only if changes, otherwise keep the fbo in memory) as well as allowing to snapshot a specific GL.Node without need to redraw it (just takes what's in the fbo). But still, I know there might be some usecases where you want to make fbos more global, especially when you have lot of concurrent surfaces (in the native case). I was considering a `framebufferGroup` concept, where you would set a number to make a sharable fbo across many nodes... anyway, that's complexifying the implementation, so i'll see later.*

**Comment by morukutsu** [comment]
Oh great, then I can't wait to see what you will come up with.
I agree, the easier & clean way is not to reuse FBOs. I like the framebufferGroup idea though!
In most use cases, it shouldn't use too much VRAM ;)

**Comment by gre** [comment]
This is now solved in gl-react v3 ! It's a complete rewrite.
- see release article: http://greweb.me/2016/12/gl-react-v3/
- see cookbook/api: https://gl-react-cookbook.surge.sh
- gl-react v3 is still in development. (hosted on: http://github.com/gre/gl-react )
- gl-react v2 is still the stable version (hosted on: https://github.com/ProjectSeptemberInc/gl-react )


Relevant example that show this issue is fixed: https://gl-react-cookbook.surge.sh/golrot and many others
